internal/dataplane/kongstate/kongstate_test.go

```diff
@@ -1475,3 +1519,131 @@ func TestFillOverrides_ServiceFailures(t *testing.T) {
+	}
+	testCases := []struct {
+		name                        string
+		initialState                *KongState
+		customEntities              []*kongv1alpha1.KongCustomEntity
+		plugins                     []*kongv1.KongPlugin
+		schemas                     map[string]kong.Schema
+		expectedCustomEntities      map[string][]custom.Object
+		expectedTranslationFailures map[k8stypes.NamespacedName]string
+	}{
+		{
+			name:         "single custom entity",
+			initialState: &KongState{},
+			customEntities: []*kongv1alpha1.KongCustomEntity{
+				{
+					TypeMeta: customEntityTypeMeta,
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "default",
+						Name:      "session-foo",
+					},
+					Spec: kongv1alpha1.KongCustomEntitySpec{
+						EntityType:     "sessions",
+						ControllerName: annotations.DefaultIngressClass,
+						Fields: apiextensionsv1.JSON{
+							Raw: []byte(`{"name":"session1"}`),
+						},
+					},
+				},
+			},
+			schemas: map[string]kong.Schema{
+				"sessions": {
+					"type":     "string",
+					"required": true,
+				},
+			},
+			expectedCustomEntities: map[string][]custom.Object{
+				"sessions": {
+					{
+						"name": "session1",
+					},
+				},
+			},
+		},
+		{
+			name:         "custom entity with unknown type",
+			initialState: &KongState{},
+			customEntities: []*kongv1alpha1.KongCustomEntity{
+				{
+					TypeMeta: customEntityTypeMeta,
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "default",
+						Name:      "session-foo",
+					},
+					Spec: kongv1alpha1.KongCustomEntitySpec{
+						EntityType:     "sessions",
+						ControllerName: annotations.DefaultIngressClass,
+						Fields: apiextensionsv1.JSON{
+							Raw: []byte(`{"name":"session1"}`),
+						},
+					},
+				},
+			},
+			expectedTranslationFailures: map[k8stypes.NamespacedName]string{
+				{
+					Namespace: "default",
+					Name:      "session-foo",
+				}: "failed to fetch entity schema for entity type sessions: schema not found",
+			},
+		},
+		// TODO: add cases including a custom entity referring to another service
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			s, err := store.NewFakeStore(store.FakeObjects{
+				KongCustomEntities: tc.customEntities,
+				KongPlugins:        tc.plugins,
+			})
+			require.NoError(t, err)
+			failuresCollector := failures.NewResourceFailuresCollector(logr.Discard())
+
+			ks := tc.initialState
+			ks.FillCustomEntities(
+				logr.Discard(), s,
+				failuresCollector,
+				&fakeSchemaGetter{schemas: tc.schemas}, "",
+			)
+			for entityType, expectedObjectList := range tc.expectedCustomEntities {
+				require.NotNil(t, ks.CustomEntities[entityType])
+				objectList := lo.Map(ks.CustomEntities[entityType].Entities, func(e CustomEntity, _ int) custom.Object {
+					return e.Object
+				})
+				require.Equal(t, expectedObjectList, objectList)
+			}
+
+			translationFailures := failuresCollector.PopResourceFailures()
+			fmt.Println(translationFailures)
```

Comment:
Do we want to keep this print? If so then let's use `t.Logf()` and described what and why we're printing it.

Reply:
removed.